src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/ProjectImports/ImportTreeCommandGroupHandler.cs

```diff
@@ -0,0 +1,244 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+using IOleServiceProvider = Microsoft.VisualStudio.OLE.Interop.IServiceProvider;
+using static Microsoft.VisualStudio.VSConstants;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.ProjectImports
+{
+    /// <summary>
+    /// Handles opening of files displayed in the project imports tree.
+    /// </summary>
+    internal abstract class ProjectImportsCommandGroupHandlerBase : IAsyncCommandGroupHandler
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IVsUIService<IVsUIShellOpenDocument> _uiShellOpenDocument;
+        private readonly IVsUIService<IVsExternalFilesManager> _externalFilesManager;
+        private readonly IVsUIService<IOleServiceProvider> _oleServiceProvider;
+
+        protected ProjectImportsCommandGroupHandlerBase(
+            IServiceProvider serviceProvider,
+            ConfiguredProject configuredProject,
+            IVsUIService<IVsUIShellOpenDocument> uiShellOpenDocument,
+            IVsUIService<IVsExternalFilesManager> externalFilesManager,
+            IVsUIService<IOleServiceProvider> oleServiceProvider)
+        {
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+            Requires.NotNull(configuredProject, nameof(configuredProject));
+            Requires.NotNull(uiShellOpenDocument, nameof(uiShellOpenDocument));
+            Requires.NotNull(externalFilesManager, nameof(externalFilesManager));
+            Requires.NotNull(oleServiceProvider, nameof(oleServiceProvider));
+
+            _serviceProvider = serviceProvider;
+            _configuredProject = configuredProject;
+            _uiShellOpenDocument = uiShellOpenDocument;
+            _externalFilesManager = externalFilesManager;
+            _oleServiceProvider = oleServiceProvider;
+        }
+
+        protected abstract bool IsOpenCommand(long commandId);
+
+        protected abstract bool IsOpenWithCommand(long commandId);
```

Comment:
Subjective, but it took me a while to understand the API here. IsOpenCommand is used to determine if the command should be handled, and IsOpenWith is only used for picking a view. it would be more logical to me if `IsOpenCommand` was renamed to `CanHandleCommand` and `IsOpenWithCommand` was changed to `GetLogicalViewForCommand(commandId`)

Reply:
Future extensions can add more `Is*Command` methods to be called from chained `if` blocks. As for `GetLogicalViewForCommand` I prefer to keep the logic internal, and using a `bool` keeps this abstraction tighter than a `Guid`.